Your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=27535&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/27535/logs/71), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/27535/logs/71)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-04-26T14:05:50.2019921Z ========================== Starting Command Output ===========================
2020-04-26T14:05:50.2024439Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/e0ab4704-626a-4ef0-8020-55314308c406.sh
2020-04-26T14:05:50.2024858Z 
2020-04-26T14:05:50.2029208Z ##[section]Finishing: Disable git automatic line ending conversion
2020-04-26T14:05:50.2046099Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/71321/merge to s
2020-04-26T14:05:50.2048999Z Task         : Get sources
2020-04-26T14:05:50.2049262Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-04-26T14:05:50.2049502Z Version      : 1.0.0
2020-04-26T14:05:50.2049667Z Author       : Microsoft
---
2020-04-26T14:05:51.4310550Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-04-26T14:05:51.4320038Z ##[command]git config gc.auto 0
2020-04-26T14:05:51.4325461Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-04-26T14:05:52.4215662Z ##[command]git config --get-all http.proxy
2020-04-26T14:05:52.4238434Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/71321/merge:refs/remotes/pull/71321/merge
---
2020-04-26T14:09:26.8244988Z  ---> f7353ccad5b1
2020-04-26T14:09:26.8245203Z Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
2020-04-26T14:09:26.8245517Z  ---> Using cache
2020-04-26T14:09:26.8245791Z  ---> ed38efbaa060
2020-04-26T14:09:26.8246892Z Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            /scripts/validate-toolstate.sh
2020-04-26T14:09:26.8247966Z  ---> c5008ef7ae8e
2020-04-26T14:09:26.8248132Z Successfully built c5008ef7ae8e
2020-04-26T14:09:26.8248469Z Successfully tagged rust-ci:latest
2020-04-26T14:09:26.8731626Z Built container sha256:c5008ef7ae8e94d7ef502e3cef26e61208e14ebdb36913f3a8bb86291bd6430b
2020-04-26T14:09:26.8731626Z Built container sha256:c5008ef7ae8e94d7ef502e3cef26e61208e14ebdb36913f3a8bb86291bd6430b
2020-04-26T14:09:26.8745356Z Looks like docker image is the same as before, not uploading
2020-04-26T14:09:35.8511915Z [CI_JOB_NAME=mingw-check]
2020-04-26T14:09:35.8747766Z [CI_JOB_NAME=mingw-check]
2020-04-26T14:09:35.8796947Z == clock drift check ==
2020-04-26T14:09:35.8806405Z   local time: Sun Apr 26 14:09:35 UTC 2020
2020-04-26T14:09:36.0151456Z   network time: Sun, 26 Apr 2020 14:09:36 GMT
2020-04-26T14:09:36.0178637Z Starting sccache server...
2020-04-26T14:09:36.1216629Z configure: processing command line
2020-04-26T14:09:36.1216927Z configure: 
2020-04-26T14:09:36.1217802Z configure: rust.parallel-compiler := True
---
2020-04-26T14:11:32.5909108Z     Checking alloc v0.0.0 (/checkout/src/liballoc)
2020-04-26T14:11:33.0238775Z error: cannot specialize on trait `core::marker::Copy`
2020-04-26T14:11:33.0239456Z     --> src/liballoc/rc.rs:1096:1
2020-04-26T14:11:33.0239852Z      |
2020-04-26T14:11:33.0240651Z 1096 | / impl<T: Copy> RcFromSlice<T> for Rc<[T]> {
2020-04-26T14:11:33.0241656Z 1097 | |     #[inline]
2020-04-26T14:11:33.0242572Z 1098 | |     fn from_slice(v: &[T]) -> Self {
2020-04-26T14:11:33.0243544Z 1099 | |         unsafe { Rc::copy_from_slice(v) }
2020-04-26T14:11:33.0245240Z 1101 | | }
2020-04-26T14:11:33.0245716Z      | |_^
2020-04-26T14:11:33.0245879Z 
2020-04-26T14:11:33.0250167Z error: cannot specialize on trait `rc::MarkerEq`
2020-04-26T14:11:33.0250167Z error: cannot specialize on trait `rc::MarkerEq`
2020-04-26T14:11:33.0251255Z     --> src/liballoc/rc.rs:1238:1
2020-04-26T14:11:33.0251938Z      |
2020-04-26T14:11:33.0252731Z 1238 | / impl<T: ?Sized + MarkerEq> RcEqIdent<T> for Rc<T> {
2020-04-26T14:11:33.0253592Z 1239 | |     #[inline]
2020-04-26T14:11:33.0256045Z 1240 | |     fn eq(&self, other: &Rc<T>) -> bool {
2020-04-26T14:11:33.0261280Z 1241 | |         Rc::ptr_eq(self, other) || **self == **other
2020-04-26T14:11:33.0275231Z 1247 | |     }
2020-04-26T14:11:33.0276020Z 1248 | | }
2020-04-26T14:11:33.0276698Z      | |_^
2020-04-26T14:11:33.0277058Z 
2020-04-26T14:11:33.0277058Z 
2020-04-26T14:11:33.0277693Z error: cannot specialize on trait `core::iter::TrustedLen`
2020-04-26T14:11:33.0278403Z     --> src/liballoc/rc.rs:1572:1
2020-04-26T14:11:33.0278973Z      |
2020-04-26T14:11:33.0279785Z 1572 | / impl<T, I: iter::TrustedLen<Item = T>> ToRcSlice<T> for I {
2020-04-26T14:11:33.0281384Z 1573 | |     fn to_rc_slice(self) -> Rc<[T]> {
2020-04-26T14:11:33.0286535Z 1574 | |         // This is the case for a `TrustedLen` iterator.
2020-04-26T14:11:33.0294505Z 1575 | |         let (low, high) = self.size_hint();
2020-04-26T14:11:33.0307873Z 1592 | |     }
2020-04-26T14:11:33.0309286Z 1593 | | }
2020-04-26T14:11:33.0313461Z      | |_^
2020-04-26T14:11:33.0322189Z 
2020-04-26T14:11:33.0322189Z 
2020-04-26T14:11:33.0323088Z error: cannot specialize on trait `core::marker::Copy`
2020-04-26T14:11:33.0324721Z    --> src/liballoc/sync.rs:976:1
2020-04-26T14:11:33.0325128Z     |
2020-04-26T14:11:33.0325733Z 976 | / impl<T: Copy> ArcFromSlice<T> for Arc<[T]> {
2020-04-26T14:11:33.0326382Z 977 | |     #[inline]
2020-04-26T14:11:33.0327072Z 978 | |     fn from_slice(v: &[T]) -> Self {
2020-04-26T14:11:33.0327859Z 979 | |         unsafe { Arc::copy_from_slice(v) }
2020-04-26T14:11:33.0329086Z 981 | | }
2020-04-26T14:11:33.0329551Z     | |_^
2020-04-26T14:11:33.0329712Z 
2020-04-26T14:11:33.0330125Z error: cannot specialize on trait `rc::MarkerEq`
2020-04-26T14:11:33.0330125Z error: cannot specialize on trait `rc::MarkerEq`
2020-04-26T14:11:33.0330653Z     --> src/liballoc/sync.rs:1782:1
2020-04-26T14:11:33.0331466Z      |
2020-04-26T14:11:33.0332094Z 1782 | / impl<T: ?Sized + crate::rc::MarkerEq> ArcEqIdent<T> for Arc<T> {
2020-04-26T14:11:33.0332793Z 1783 | |     #[inline]
2020-04-26T14:11:33.0333488Z 1784 | |     fn eq(&self, other: &Arc<T>) -> bool {
2020-04-26T14:11:33.0334313Z 1785 | |         Arc::ptr_eq(self, other) || **self == **other
2020-04-26T14:11:33.0335523Z 1791 | |     }
2020-04-26T14:11:33.0336275Z 1792 | | }
2020-04-26T14:11:33.0336919Z      | |_^
2020-04-26T14:11:33.0337105Z 
2020-04-26T14:11:33.0337105Z 
2020-04-26T14:11:33.0337614Z error: cannot specialize on trait `core::iter::TrustedLen`
2020-04-26T14:11:33.0338212Z     --> src/liballoc/sync.rs:2123:1
2020-04-26T14:11:33.0338652Z      |
2020-04-26T14:11:33.0339689Z 2123 | / impl<T, I: iter::TrustedLen<Item = T>> ToArcSlice<T> for I {
2020-04-26T14:11:33.0340472Z 2124 | |     fn to_arc_slice(self) -> Arc<[T]> {
2020-04-26T14:11:33.0341335Z 2125 | |         // This is the case for a `TrustedLen` iterator.
2020-04-26T14:11:33.0342256Z 2126 | |         let (low, high) = self.size_hint();
2020-04-26T14:11:33.0343564Z 2143 | |     }
2020-04-26T14:11:33.0344143Z 2144 | | }
2020-04-26T14:11:33.0344632Z      | |_^
2020-04-26T14:11:33.0344795Z 
2020-04-26T14:11:33.0344795Z 
2020-04-26T14:11:33.0345242Z error: cannot specialize on trait `vec::IsZero`
2020-04-26T14:11:33.0345847Z     --> src/liballoc/vec.rs:1816:1
2020-04-26T14:11:33.0346218Z      |
2020-04-26T14:11:33.0346789Z 1816 | / impl<T: Clone + IsZero> SpecFromElem for T {
2020-04-26T14:11:33.0347416Z 1817 | |     #[inline]
2020-04-26T14:11:33.0348112Z 1818 | |     fn from_elem(elem: T, n: usize) -> Vec<T> {
2020-04-26T14:11:33.0348798Z 1819 | |         if elem.is_zero() {
2020-04-26T14:11:33.0349914Z 1825 | |     }
2020-04-26T14:11:33.0350601Z 1826 | | }
2020-04-26T14:11:33.0351111Z      | |_^
2020-04-26T14:11:33.0351298Z 
2020-04-26T14:11:33.0351298Z 
2020-04-26T14:11:33.0386076Z error: cannot specialize on trait `core::iter::TrustedLen`
2020-04-26T14:11:33.0386771Z     --> src/liballoc/vec.rs:2087:1
2020-04-26T14:11:33.0387190Z      |
2020-04-26T14:11:33.0387764Z 2087 | / impl<T, I> SpecExtend<T, I> for Vec<T>
2020-04-26T14:11:33.0388398Z 2088 | | where
2020-04-26T14:11:33.0389072Z 2089 | |     I: TrustedLen<Item = T>,
2020-04-26T14:11:33.0389690Z 2090 | | {
2020-04-26T14:11:33.0390684Z 2123 | |     }
2020-04-26T14:11:33.0391231Z 2124 | | }
2020-04-26T14:11:33.0391710Z      | |_^
2020-04-26T14:11:33.0391872Z 
2020-04-26T14:11:33.0391872Z 
2020-04-26T14:11:33.0392293Z error: cannot specialize on trait `core::marker::Copy`
2020-04-26T14:11:33.0392977Z     --> src/liballoc/vec.rs:2165:1
2020-04-26T14:11:33.0393385Z      |
2020-04-26T14:11:33.0394005Z 2165 | / impl<'a, T: 'a> SpecExtend<&'a T, slice::Iter<'a, T>> for Vec<T>
2020-04-26T14:11:33.0394678Z 2166 | | where
2020-04-26T14:11:33.0396002Z 2168 | | {
2020-04-26T14:11:33.0396482Z ...    |
2020-04-26T14:11:33.0396982Z 2178 | |     }
2020-04-26T14:11:33.0397529Z 2179 | | }
---
2020-04-26T14:11:33.0453011Z error: could not compile `alloc`.
2020-04-26T14:11:33.0462130Z 
2020-04-26T14:11:33.0462720Z To learn more, run the command again with --verbose.
2020-04-26T14:11:33.0463456Z warning: build failed, waiting for other jobs to finish...
2020-04-26T14:11:33.5258558Z {"reason":"build-finished","success":false}
2020-04-26T14:11:33.5346905Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/src/libtest/Cargo.toml" "--message-format" "json-render-diagnostics"
2020-04-26T14:11:33.5347599Z expected success, got: exit code: 101
2020-04-26T14:11:33.5357249Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
2020-04-26T14:11:33.5357577Z Build completed unsuccessfully in 0:01:57
2020-04-26T14:11:33.5357577Z Build completed unsuccessfully in 0:01:57
2020-04-26T14:11:33.5454084Z == clock drift check ==
2020-04-26T14:11:33.5469717Z   local time: Sun Apr 26 14:11:33 UTC 2020
2020-04-26T14:11:33.8120506Z   network time: Sun, 26 Apr 2020 14:11:33 GMT
2020-04-26T14:11:35.0604916Z 
2020-04-26T14:11:35.0604916Z 
2020-04-26T14:11:35.0671560Z ##[error]Bash exited with code '1'.
2020-04-26T14:11:35.0685054Z ##[section]Finishing: Run build
2020-04-26T14:11:35.0734153Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/71321/merge to s
2020-04-26T14:11:35.0739230Z Task         : Get sources
2020-04-26T14:11:35.0739529Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-04-26T14:11:35.0739796Z Version      : 1.0.0
2020-04-26T14:11:35.0740115Z Author       : Microsoft
2020-04-26T14:11:35.0740115Z Author       : Microsoft
2020-04-26T14:11:35.0740407Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2020-04-26T14:11:35.0740729Z ==============================================================================
2020-04-26T14:11:35.4098771Z Cleaning any cached credential from repository: rust-lang/rust (GitHub)
2020-04-26T14:11:35.4150671Z ##[section]Finishing: Checkout rust-lang/rust@refs/pull/71321/merge to s
2020-04-26T14:11:35.4236174Z Cleaning up task key
2020-04-26T14:11:35.4237305Z Start cleaning up orphan processes.
2020-04-26T14:11:35.4446466Z Terminate orphan process: pid (3504) (python)
2020-04-26T14:11:35.4636585Z ##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

--- a/src/liballoc/collections/btree/map.rs
+++ b/src/liballoc/collections/btree/map.rs
@@ -215,59 +215,6 @@ impl<K: Clone, V: Clone> Clone for BTreeMap<K, V> {
             clone_subtree(self.root.as_ref().unwrap().as_ref())
         }
     }
-
-    fn clone_from(&mut self, other: &Self) {
-        BTreeClone::clone_from(self, other);
-    }
-}
-
-trait BTreeClone {
-    fn clone_from(&mut self, other: &Self);
-}
-
-impl<K: Clone, V: Clone> BTreeClone for BTreeMap<K, V> {
-    default fn clone_from(&mut self, other: &Self) {
-        *self = other.clone();
-    }
-}
-
-impl<K: Clone + Ord, V: Clone> BTreeClone for BTreeMap<K, V> {
-    fn clone_from(&mut self, other: &Self) {
-        // This truncates `self` to `other.len()` by calling `split_off` on
-        // the first key after `other.len()` elements if it exists.
-        let split_off_key = if self.len() > other.len() {
-            let diff = self.len() - other.len();
-            if diff <= other.len() {
-                self.iter().nth_back(diff - 1).map(|pair| (*pair.0).clone())
-            } else {
-                self.iter().nth(other.len()).map(|pair| (*pair.0).clone())
-            }
-        } else {
-            None
-        };
-        if let Some(key) = split_off_key {
-            self.split_off(&key);
-        }
-
-        let mut siter = self.range_mut(..);
-        let mut oiter = other.iter();
-        // After truncation, `self` is at most as long as `other` so this loop
-        // replaces every key-value pair in `self`. Since `oiter` is in sorted
-        // order and the structure of the `BTreeMap` stays the same,
-        // the BTree invariants are maintained at the end of the loop.
-        while !siter.is_empty() {
-            if let Some((ok, ov)) = oiter.next() {
-                // SAFETY: This is safe because `siter` is nonempty.
-                let (sk, sv) = unsafe { siter.next_unchecked() };
-                sk.clone_from(ok);
-                sv.clone_from(ov);
-            } else {
-                break;
-            }
-        }
-        // If `other` is longer than `self`, the remaining elements are inserted.
-        self.extend(oiter.map(|(k, v)| ((*k).clone(), (*v).clone())));
-    }
 }
 
 impl<K, Q: ?Sized> super::Recover<Q> for BTreeMap<K, ()>

--- a/src/liballoc/lib.rs
+++ b/src/liballoc/lib.rs
@@ -108,7 +108,7 @@
 #![feature(ptr_offset_from)]
 #![feature(rustc_attrs)]
 #![feature(receiver_trait)]
-#![feature(specialization)]
+#![feature(min_specialization)]
 #![feature(staged_api)]
 #![feature(std_internals)]
 #![feature(str_internals)]

--- a/src/liballoc/rc.rs
+++ b/src/liballoc/rc.rs
@@ -249,7 +249,7 @@ use core::mem::{self, align_of, align_of_val, forget, size_of_val};
 use core::ops::{CoerceUnsized, Deref, DispatchFromDyn, Receiver};
 use core::pin::Pin;
 use core::ptr::{self, NonNull};
-use core::slice::{self, from_raw_parts_mut};
+use core::slice::from_raw_parts_mut;
 
 use crate::alloc::{box_free, handle_alloc_error, AllocInit, AllocRef, Global, Layout};
 use crate::string::String;
@@ -1221,6 +1221,12 @@ impl<T: ?Sized + PartialEq> RcEqIdent<T> for Rc<T> {
     }
 }
 
+// Hack to allow specializing on `Eq` even though `Eq` has a method.
+#[rustc_unsafe_specialization_marker]
+pub(crate) trait MarkerEq: PartialEq<Self> {}
+
+impl<T: Eq> MarkerEq for T {}
+
 /// We're doing this specialization here, and not as a more general optimization on `&T`, because it
 /// would otherwise add a cost to all equality checks on refs. We assume that `Rc`s are used to
 /// store large values, that are slow to clone, but also heavy to check for equality, causing this
@@ -1229,7 +1235,7 @@ impl<T: ?Sized + PartialEq> RcEqIdent<T> for Rc<T> {
 ///
 /// We can only do this when `T: Eq` as a `PartialEq` might be deliberately irreflexive.
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<T: ?Sized + Eq> RcEqIdent<T> for Rc<T> {
+impl<T: ?Sized + MarkerEq> RcEqIdent<T> for Rc<T> {
     #[inline]
     fn eq(&self, other: &Rc<T>) -> bool {
         Rc::ptr_eq(self, other) || **self == **other
@@ -1548,25 +1554,25 @@ impl<T> iter::FromIterator<T> for Rc<[T]> {
     /// # assert_eq!(&*evens, &*(0..10).collect::<Vec<_>>());
     /// ```
     fn from_iter<I: iter::IntoIterator<Item = T>>(iter: I) -> Self {
-        RcFromIter::from_iter(iter.into_iter())
+        ToRcSlice::to_rc_slice(iter.into_iter())
     }
 }
 
 /// Specialization trait used for collecting into `Rc<[T]>`.
-trait RcFromIter<T, I> {
-    fn from_iter(iter: I) -> Self;
+trait ToRcSlice<T>: Iterator<Item = T> + Sized {
+    fn to_rc_slice(self) -> Rc<[T]>;
 }
 
-impl<T, I: Iterator<Item = T>> RcFromIter<T, I> for Rc<[T]> {
-    default fn from_iter(iter: I) -> Self {
-        iter.collect::<Vec<T>>().into()
+impl<T, I: Iterator<Item = T>> ToRcSlice<T> for I {
+    default fn to_rc_slice(self) -> Rc<[T]> {
+        self.collect::<Vec<T>>().into()
     }
 }
 
-impl<T, I: iter::TrustedLen<Item = T>> RcFromIter<T, I> for Rc<[T]> {
-    default fn from_iter(iter: I) -> Self {
+impl<T, I: iter::TrustedLen<Item = T>> ToRcSlice<T> for I {
+    fn to_rc_slice(self) -> Rc<[T]> {
         // This is the case for a `TrustedLen` iterator.
-        let (low, high) = iter.size_hint();
+        let (low, high) = self.size_hint();
         if let Some(high) = high {
             debug_assert_eq!(
                 low,
@@ -1577,26 +1583,12 @@ impl<T, I: iter::TrustedLen<Item = T>> RcFromIter<T, I> for Rc<[T]> {
 
             unsafe {
                 // SAFETY: We need to ensure that the iterator has an exact length and we have.
-                Rc::from_iter_exact(iter, low)
+                Rc::from_iter_exact(self, low)
             }
         } else {
             // Fall back to normal implementation.
-            iter.collect::<Vec<T>>().into()
+            self.collect::<Vec<T>>().into()
         }
-    }
-}
-
-impl<'a, T: 'a + Clone> RcFromIter<&'a T, slice::Iter<'a, T>> for Rc<[T]> {
-    fn from_iter(iter: slice::Iter<'a, T>) -> Self {
-        // Delegate to `impl<T: Clone> From<&[T]> for Rc<[T]>`.
-        //
-        // In the case that `T: Copy`, we get to use `ptr::copy_nonoverlapping`
-        // which is even more performant.
-        //
-        // In the fall-back case we have `T: Clone`. This is still better
-        // than the `TrustedLen` implementation as slices have a known length
-        // and so we get to avoid calling `size_hint` and avoid the branching.
-        iter.as_slice().into()
     }
 }
 

--- a/src/liballoc/sync.rs
+++ b/src/liballoc/sync.rs
@@ -20,7 +20,7 @@ use core::mem::{self, align_of, align_of_val, size_of_val};
 use core::ops::{CoerceUnsized, Deref, DispatchFromDyn, Receiver};
 use core::pin::Pin;
 use core::ptr::{self, NonNull};
-use core::slice::{self, from_raw_parts_mut};
+use core::slice::from_raw_parts_mut;
 use core::sync::atomic;
 use core::sync::atomic::Ordering::{Acquire, Relaxed, Release, SeqCst};
 
@@ -1779,7 +1779,7 @@ impl<T: ?Sized + PartialEq> ArcEqIdent<T> for Arc<T> {
 ///
 /// We can only do this when `T: Eq` as a `PartialEq` might be deliberately irreflexive.
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<T: ?Sized + Eq> ArcEqIdent<T> for Arc<T> {
+impl<T: ?Sized + crate::rc::MarkerEq> ArcEqIdent<T> for Arc<T> {
     #[inline]
     fn eq(&self, other: &Arc<T>) -> bool {
         Arc::ptr_eq(self, other) || **self == **other
@@ -2105,25 +2105,25 @@ impl<T> iter::FromIterator<T> for Arc<[T]> {
     /// # assert_eq!(&*evens, &*(0..10).collect::<Vec<_>>());
     /// ```
     fn from_iter<I: iter::IntoIterator<Item = T>>(iter: I) -> Self {
-        ArcFromIter::from_iter(iter.into_iter())
+        ToArcSlice::to_arc_slice(iter.into_iter())
     }
 }
 
 /// Specialization trait used for collecting into `Arc<[T]>`.
-trait ArcFromIter<T, I> {
-    fn from_iter(iter: I) -> Self;
+trait ToArcSlice<T>: Iterator<Item = T> + Sized {
+    fn to_arc_slice(self) -> Arc<[T]>;
 }
 
-impl<T, I: Iterator<Item = T>> ArcFromIter<T, I> for Arc<[T]> {
-    default fn from_iter(iter: I) -> Self {
-        iter.collect::<Vec<T>>().into()
+impl<T, I: Iterator<Item = T>> ToArcSlice<T> for I {
+    default fn to_arc_slice(self) -> Arc<[T]> {
+        self.collect::<Vec<T>>().into()
     }
 }
 
-impl<T, I: iter::TrustedLen<Item = T>> ArcFromIter<T, I> for Arc<[T]> {
-    default fn from_iter(iter: I) -> Self {
+impl<T, I: iter::TrustedLen<Item = T>> ToArcSlice<T> for I {
+    fn to_arc_slice(self) -> Arc<[T]> {
         // This is the case for a `TrustedLen` iterator.
-        let (low, high) = iter.size_hint();
+        let (low, high) = self.size_hint();
         if let Some(high) = high {
             debug_assert_eq!(
                 low,
@@ -2134,26 +2134,12 @@ impl<T, I: iter::TrustedLen<Item = T>> ArcFromIter<T, I> for Arc<[T]> {
 
             unsafe {
                 // SAFETY: We need to ensure that the iterator has an exact length and we have.
-                Arc::from_iter_exact(iter, low)
+                Arc::from_iter_exact(self, low)
             }
         } else {
             // Fall back to normal implementation.
-            iter.collect::<Vec<T>>().into()
+            self.collect::<Vec<T>>().into()
         }
-    }
-}
-
-impl<'a, T: 'a + Clone> ArcFromIter<&'a T, slice::Iter<'a, T>> for Arc<[T]> {
-    fn from_iter(iter: slice::Iter<'a, T>) -> Self {
-        // Delegate to `impl<T: Clone> From<&[T]> for Arc<[T]>`.
-        //
-        // In the case that `T: Copy`, we get to use `ptr::copy_nonoverlapping`
-        // which is even more performant.
-        //
-        // In the fall-back case we have `T: Clone`. This is still better
-        // than the `TrustedLen` implementation as slices have a known length
-        // and so we get to avoid calling `size_hint` and avoid the branching.
-        iter.as_slice().into()
     }
 }
 

--- a/src/liballoc/vec.rs
+++ b/src/liballoc/vec.rs
@@ -1619,8 +1619,8 @@ impl<T: Default> Vec<T> {
     #[unstable(feature = "vec_resize_default", issue = "41758")]
     #[rustc_deprecated(
         reason = "This is moving towards being removed in favor \
-        of `.resize_with(Default::default)`.  If you disagree, please comment \
-        in the tracking issue.",
+                  of `.resize_with(Default::default)`.  If you disagree, please comment \
+                  in the tracking issue.",
         since = "1.33.0"
     )]
     pub fn resize_default(&mut self, new_len: usize) {
@@ -1825,6 +1825,7 @@ impl<T: Clone + IsZero> SpecFromElem for T {
     }
 }
 
+#[rustc_specialization_trait]
 unsafe trait IsZero {
     /// Whether this value is zero
     fn is_zero(&self) -> bool;
@@ -1874,18 +1875,14 @@ unsafe impl<T> IsZero for *mut T {
     }
 }
 
-// `Option<&T>`, `Option<&mut T>` and `Option<Box<T>>` are guaranteed to represent `None` as null.
-// For fat pointers, the bytes that would be the pointer metadata in the `Some` variant
-// are padding in the `None` variant, so ignoring them and zero-initializing instead is ok.
+// `Option<&T>` and `Option<Box<T>>` are guaranteed to represent `None` as null.
+// For fat pointers, the bytes that would be the pointer metadata in the `Some`
+// variant are padding in the `None` variant, so ignoring them and
+// zero-initializing instead is ok.
+// `Option<&mut T>` never implements `Clone`, so there's no need for an impl of
+// `SpecFromElem`.
 
 unsafe impl<T: ?Sized> IsZero for Option<&T> {
-    #[inline]
-    fn is_zero(&self) -> bool {
-        self.is_none()
-    }
-}
-
-unsafe impl<T: ?Sized> IsZero for Option<&mut T> {
     #[inline]
     fn is_zero(&self) -> bool {
         self.is_none()

--- a/src/libcore/iter/traits/marker.rs
+++ b/src/libcore/iter/traits/marker.rs
@@ -13,6 +13,7 @@
 /// [`Iterator::fuse`]: ../../std/iter/trait.Iterator.html#method.fuse
 /// [`Fuse`]: ../../std/iter/struct.Fuse.html
 #[stable(feature = "fused", since = "1.26.0")]
+#[rustc_unsafe_specialization_marker]
 pub trait FusedIterator: Iterator {}
 
 #[stable(feature = "fused", since = "1.26.0")]
@@ -38,6 +39,7 @@ impl<I: FusedIterator + ?Sized> FusedIterator for &mut I {}
 /// [`usize::MAX`]: ../../std/usize/constant.MAX.html
 /// [`.size_hint`]: ../../std/iter/trait.Iterator.html#method.size_hint
 #[unstable(feature = "trusted_len", issue = "37572")]
+#[rustc_unsafe_specialization_marker]
 pub unsafe trait TrustedLen: Iterator {}
 
 #[unstable(feature = "trusted_len", issue = "37572")]

--- a/src/libcore/marker.rs
+++ b/src/libcore/marker.rs
@@ -363,6 +363,13 @@ pub trait StructuralEq {
 /// [impls]: #implementors
 #[stable(feature = "rust1", since = "1.0.0")]
 #[lang = "copy"]
+// FIXME(matthewjasper) This allows copying a type that doesn't implement
+// `Copy` because of unsatisfied lifetime bounds (copying `A<'_>` when only
+// `A<'static>: Copy` and `A<'_>: Clone`).
+// We have this attribute here for now only because there are quite a few
+// existing specializations on `Copy` that already exist in the standard
+// library, and there's no way to safely have this behavior right now.
+#[rustc_unsafe_specialization_marker]
 pub trait Copy: Clone {
     // Empty.
 }

